### PR TITLE
Allow DWARF with source map

### LINKED ISSUE #16699: Support generation of both DWARF and source maps in a single invocation
Currently you can build either with DWARF or source maps debugging information, but not both. But this limitation is artificial (since the source map is actually generated from a DWARF line table), so it should be straightforward to just not strip the DWARF after generating the source map.  One possible use case is to have full dwarf info for part of the binary, and source maps for all of it.

### PR DESCRIPTION
**Pull request by aheejin**
This allows generation of DWARF info and source map at the same time.
This adds an internal setting, `GENERATE_DWARF`, which is a counterpart
of existing `GENERATE_SORUCE_MAP`, because with only `DEBUG_LEVEL` we
now don't have a way to tell whether DWARF info is explicitly requested
or not, because we set `DEBUG_LEVEL` to 3 when source map is requested.

This adds some simple utility functions to `test_other.py` for DWARF and
source map verification. Also this enables `try_delete` to delete a list
of files.

Fixes #16699.

### CODE REVIEW

File: emcc.py [hunk 1]
```diff
@@ -3169,6 +3169,8 @@ def consume_arg_file():
         else:
           # for 3+, report -g3 to clang as -g4 etc. are not accepted
           newargs[i] = '-g3'
+          if settings.DEBUG_LEVEL == 3:
+            settings.GENERATE_DWARF = 1
```
Comment (kripken) [inline emcc.py:3173]: Now that this exists, maybe a nice followup PR could use it in more places. I think we have things like `DEBUG_LEVEL == 3` in various places.

**Review by kripken — APPROVED** [review]
lgtm and nice refactoring!

Please add an entry to the changelog for this.

File: emcc.py (hunk 1 above)
Comment (aheejin) [reply emcc.py:3173]: Good idea, will check for that.

**Review by dschuff — APPROVED** [review]

File: tools/tempfiles.py [hunk 2]
```diff
@@ -11,9 +11,17 @@
 import sys
 
 
+def try_delete(pathnames):
+  if isinstance(pathnames, list):
+    for pathname in pathnames:
+      try_delete_one(pathname)
+  else:
+    try_delete_one(pathnames)
```
Comment (sbc100) [inline tools/tempfiles.py:19]: I'd rather not add helpers like this that are only used in test code.   Also I'm not a fan of this type of polymorphism in python.    How about we have a new helper to in `test_other.py` called `delete_files`?

File: tests/test_other.py [hunk 3]
```diff
@@ -2425,20 +2425,59 @@ def compile_to_O0_executable(compile_args):
...
+    source_file = 'hello_world.cpp'
+    js_file = 'a.out.js'
+    wasm_file = 'a.out.wasm'
+    map_file = 'a.out.wasm.map'
+
+    # Generate only DWARF
+    self.emcc(test_file(source_file), ['-g'], js_file)
+    self.verify_dwarf_exists(wasm_file)
+    self.assertFalse(os.path.isfile(map_file))
+    try_delete([wasm_file, map_file, js_file])
```
Comment (sbc100) [inline tests/test_other.py:2469]: Actually I think you can just `self.clear()` here instead and avoid the new helper completely?

Comment (aheejin) [reply tests/test_other.py:2469]: Oh, didn't know that exists. But I guess the new helper can be useful in some places?

Comment (aheejin) [reply tests/test_other.py:2469]: Used `self.clear()` in #17519.

Comment (sbc100) [reply tests/test_other.py:2469]: I make it part of a bigger refactoring in https://github.com/emscripten-core/emscripten/pull/17512.. thanks for reminding about that code I've been meaning to clean up :)

File: emcc.py (hunk 1 above)
Comment (aheejin) [reply emcc.py:3184]: Tried to search for the usage, but it's tricky. Usually when we check for `DEBUG_LEVEL == 3` or `DEBUG_LEVEL >= 3`, it's for generating/preserving DWARF info. But the tricky thing is, even if only source map is required, it needs DWARF info to be generated until the middle of the process and we only get rid of DWARF at the end after generating source map from it. So `DEBUG_LEVEL == 3` is synonymous with "Do we need DWARF info _in the beginning_?", and I'm not sure if replacing this with `if settings.GENERATE_DWARF or settings.GENERATE_SOURCE_MAP` will be much of a improvement. If we want to go that route, I think it's better to redefine the meaning of `DEBUG_LEVEL` or create separate variables for other `DEBUG_LEVEL` (1 and 2) too; but I'm not sure if we want to go that far.

Comment (kripken) [reply emcc.py:3184]: Makes sense, thanks for checking. Nothing to do for that, then.

**Pull request merged**